File: Semester 2/COS20007 Object Oriented Programming/Week 11/11.1P Clock In Another Language/clock.cpp
#include <iostream>
#include <iomanip>
#include <sstream>
#include <string>

// This class provides methods to increment, reset, and retrieve
class Counter {
private:
    int count;

public:
    Counter() : count(0) {}

    void increment() {
        count++;
    }

    void reset() {
        count = 0;
    }

    int getCount() const {
        return count;
    }
};

// Keep track of time
class Clock {
private:
    Counter seconds;
    Counter minutes;
    Counter hours;

public:
    Clock() {}

    void tick() {
        seconds.increment();
        if (seconds.getCount() > 59) {
            minutes.increment();
            seconds.reset();
        }
        if (minutes.getCount() > 59) {
            hours.increment();
            minutes.reset();
        }
        if (hours.getCount() > 23) {
            hours.reset();
            minutes.reset();
            seconds.reset();
        }
    }

    void resetClock() {
        hours.reset();
        minutes.reset();
        seconds.reset();
    }

    //To build the formatted string representation of the time
    //Setfill manipulator ensures that leading zeros are added
    std::string getTime() const {
        std::ostringstream oss; 
        oss << std::setfill('0') << std::setw(2) << hours.getCount() << ":"
            << std::setfill('0') << std::setw(2) << minutes.getCount() << ":"
            << std::setfill('0') << std::setw(2) << seconds.getCount();
        return oss.str();
    }
};

int main() {
    Clock myClock;
    myClock.resetClock();

    for (int i = 0; i < 260; i++) {
        myClock.tick();
        std::cout << myClock.getTime() << std::endl;
    }

    return 0;
}

File: Semester 2/COS20007 Object Oriented Programming/Week 3/3.1P ClockClass/ClockClass/Clock.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace ClockClass
{
    public class Clock
    {
        private Counter _secs;
        private Counter _mins;
        private Counter _hrs;

        public Clock()
        {
            _secs = new Counter("seconds");
            _mins = new Counter("minutes");
            _hrs = new Counter("hours");
        }


        public void Tick()
        {
            _secs.Increment();
            if (_secs.Ticks > 59)
            {
                _mins.Increment();
                _secs.Reset();
            }
            if (_mins.Ticks > 59)
            {
                _hrs.Increment();
                _mins.Reset();
            }
            if (_hrs.Ticks > 23)
            {
                _hrs.Reset();
                _mins.Reset();
                _secs.Reset();
            }
        }

        public void Reset()
        {
            _hrs.Reset();
            _mins.Reset();
            _secs.Reset();
        }

        public string Time
        {
            get
            {
                return
                    $"{_hrs.Ticks:00}:{_mins.Ticks:00}:{_secs.Ticks:00}";
            }
        }
    }
}
